csvparser/services.py

```diff
@@ -0,0 +1,199 @@
+# pylint: disable=no-else-raise
+"""
+CSV Parser Service
+"""
+import csv
+import re
+
+from django.core.exceptions import ValidationError
+
+required_columns = ['username', 'email', 'password',
+                    'date_of_birth', 'phone_number', 'blood_group']
+blood_group_types = ['A+', 'B+', 'O+', 'AB+', 'AB-', 'O-', 'B-', 'A-']
+
+
+def decode_utf8(input_iterator):
+    """
+    Generator that decodes a utf-8 encoded
+    file input line by line
+    """
+    for line in input_iterator:
+        yield line if isinstance(line, str) else line.decode('utf-8')
+
+
+def validate_username(value):
+    """
+    Validates username column using regex
+    """
+    reg = re.compile(
+        r'[A-Za-z][A-Za-z0-9_]{7,29}')
+    if not reg.match(value):
+        raise ValidationError(f'{value} is not valid for column username')
+    else:
+        return True
+
+
+def validate_email(value):
+    """
+    Validates email column using regex
+    """
+    reg = re.compile(
+        r'(?=.{1,64}@)[A-Za-z0-9_-]+(\\.[A-Za-z0-9_-]+)*@[^-][A-Za-z0-9-]+(\\.'
+        '[A-Za-z0-9-]+)*(\\.[A-Za-z]{2,})')
+    if not reg.match(value):
+        raise ValidationError(f'{value} is not valid for column email')
+    else:
+        return True
+
+
+def validate_password(value):
+    """
+    Validates password column using regex
+    """
+    reg = re.compile(r'(?=.*[a-z])(?=.*[A-Z])[A-Za-z\d@$!#%*?&]{8,18}')
+    if not reg.match(value):
+        raise ValidationError(f'{value} is not valid for column password')
+    else:
+        return True
+
+
+def validate_date_of_birth(value):
+    """
+    Validates Date of birth column using regex
+    """
+    reg = re.compile(
+        '((?:19|20)[0-9][0-9])-(0?[1-9]|1[012])-(0?[1-9]|[12][0-9]|3[01])')
+    if not reg.match(value):
+        raise ValidationError(f'{value} is not valid for column date_of_birth')
+    else:
+        return True
+
+
+def validate_phone_number(value):
+    """
+    Validates phone number column using regex
+    """
+    reg = re.compile(r'\+?1?\d{9,15}')
+    if not reg.match(value):
+        raise ValidationError(f'{value} is not valid for column phone_number')
+    else:
+        return True
+
+
+def validate_blood_group(value):
+    """
+    Validates Blood grouo column
+    """
+    if value in blood_group_types:
+        return True
+    else:
+        raise ValidationError(f'{value} is not valid for column blood_group')
+
+
+class CsvParser:
+    """
+    CSV Parser Class for adding bulk donor users
+    """
+
+    def __init__(self, file):
+        self.file = file
+        self.error_messages = []
+
+    def add_errors_to_error_list(self, errors):
+        """
+        Add errors list to error_messages list
+        """
+        self.error_messages.extend(errors)
+
+    def check_file_has_content(self, file_reader):
+        """
+        Check if file has content or not
+        """
+        if file_reader.fieldnames is not None:
```

Comment:
in case of empty file we still have the reader when using DictReader but we would have fieldnames none in this case..... I was checking this condition "if file_reader is not None" earlier but it wasn't working as it was not none